homeassistant/components/bond/fan.py

```diff
@@ -0,0 +1,115 @@
+    SPEED_HIGH,
+    SPEED_LOW,
+    SPEED_MEDIUM,
+    SPEED_OFF,
+    SUPPORT_SET_SPEED,
+    FanEntity,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import Entity
+
+from ...const import ATTR_NAME
+from .const import DOMAIN
+from .utils import BondDevice, get_bond_devices
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: Callable[[List[Entity], bool], None],
+) -> None:
+    """Set up Bond fan devices."""
+    bond: Bond = hass.data[DOMAIN][entry.entry_id]
+
+    devices = await hass.async_add_executor_job(get_bond_devices, hass, bond)
+
+    fans = [
+        BondFan(bond, device)
+        for device in devices
+        if device.type == BOND_DEVICE_TYPE_CEILING_FAN
+    ]
+
+    async_add_entities(fans, True)
+
+
+class BondFan(FanEntity):
+    """Representation of a Bond fan."""
+
+    def __init__(self, bond: Bond, device: BondDevice):
+        """Create HA entity representing Bond fan."""
+        self._bond = bond
+        self._device = device
+
+        self._power: Optional[bool] = None
+        self._speed: Optional[int] = None
+
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Get unique ID for the entity."""
+        return self._device.device_id
+
+    @property
+    def name(self) -> Optional[str]:
+        """Get entity name."""
+        return self._device.name
+
+    @property
+    def device_info(self) -> Optional[Dict[str, Any]]:
+        """Get a an HA device representing this fan."""
+        return {ATTR_NAME: self.name, "identifiers": {(DOMAIN, self._device.device_id)}}
+
+    @property
+    def assumed_state(self) -> bool:
+        """Let HA know this entity relies on an assumed state tracked by Bond."""
+        return True
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        features = 0
+        if self._device.supports_command("SetSpeed"):
+            features |= SUPPORT_SET_SPEED
+        return features
+
+    @property
+    def speed(self) -> Optional[str]:
+        """Return the current speed."""
+        if self._power is None:
+            return None
+        if self._power == 0:
+            return SPEED_OFF
+
+        return self.speed_list[self._speed] if self._speed is not None else None
+
+    @property
+    def speed_list(self) -> list:
+        """Get the list of available speeds."""
+        return [SPEED_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]
+
+    def update(self):
+        """Fetch assumed state of the fan from the hub using API."""
+        state: dict = self._bond.getDeviceState(self._device.device_id)
+        self._power = state.get("power")
+        self._speed = state.get("speed")
+
+    def set_speed(self, speed: str) -> None:
+        """Set the desired speed for the fan."""
+        speed_index = self.speed_list.index(speed)
```

Comment:
For 6 speed fans, this would make `SPEED_HIGH` as 3, which would only be a medium speed.  Is there going to be consideration for 6 speed fans?

Reply:
@steve-gombos I don't own a 6-speed fan unfortunately, but I can put up a PR to add support, will make sure to add you for testing.

Reply:
I will pull this down and take a look at it this evening when I get some time.

Reply:
> For 6 speed fans, this would make `SPEED_HIGH` as 3, which would only be a medium speed. Is there going to be consideration for 6 speed fans? The existing custom component I am using for Bond tries to do a rough calculation for this based on the `max_speed` property.

@steve-gombos I have added the support for custom speeds. I have a pending PR right now on which this support is on top of. Once it is reviewed and merged I will put out a PR for the custom speeds. Meanwhile, you can preview it on my fork: https://github.com/prystupa/core/pull/1/files

Reply:
Just pulled down your changes.  Working on my fans.  Thanks for the quick update on that.

Reply:
Saw the PR was cancelled, and I understand the reasoning, there is a discussion already on the architecture repo about additional fan speeds here: https://github.com/home-assistant/architecture/issues/127  Doesn't seem to be going anywhere.

I know bond can support some fans with a crazy amount of speeds (ie 10, 7, 6, etc...)  I feel like something like this could be a compromise to make use of the available speed range, which sticking to the supported speeds in HA.

```py
import math

    def __init__(self, hub: BondHub, device: BondDevice):
        self._max_speed = self._device.props.get("max_speed", 3)
        self._speed_factor = int(math.floor(self._max_speed / 3))

    @property
    def speed(self) -> Optional[str]:
        """Return the current speed."""
        if self._power is None:
            return None
        if self._power == 0:
            return SPEED_OFF

        adjusted_speed = int(self._speed / self._speed_factor)
        return self.speed_list[adjusted_speed] if self._speed is not None else None

    def set_speed(self, speed: str) -> None:
        """Set the desired speed for the fan."""
        speed_index = self.speed_list.index(speed)
        actual_speed = int(speed_index * self._speed_factor)
        self._hub.bond.setSpeed(self._device.device_id, speed=actual_speed)
```

Reply:
@steve-gombos no worries, I am putting together a PR now that will map custom speeds to standard. So, at least you will have low == 1, medium == 3, and high == 6 for 6-speed fan

Reply:
@steve-gombos Please check this out and give it a try on your fans: https://github.com/home-assistant/core/pull/37808
This maps bond speeds to standard HA speeds as discussed

Reply:
This works.